benchmark/__init__.py

```diff
@@ -0,0 +1,610 @@
+                    subprocess.check_call([
+                        "gsutil", "-m", "-q", "rm", "-r",
+                        "{}".format(self.url_prefix)
+                    ])
+
+                try:
+                    gsutil_start = time.time()
+                    for file in tmp_dir.path.iterdir():
+                        self._gsutil_upload_files(path=file)
+                    gsutil_end = time.time()
+                finally:
+                    subprocess.check_call([
+                        "gsutil", "-m", "-q", "rm", "-r",
+                        "{}".format(self.url_prefix)
+                    ])
+
+            print_benchmark(
+                benchmark="Benchmark upload {} single files".format(testcase),
+                time_gswrap=gswrap_end - gswrap_start,
+                time_gsutil=gsutil_end - gsutil_start)
+
+    def benchmark_upload_big_files(self):
+        number_of_files = 3
+        for size in [10, 1024, 1024**2, 200 * 1024**2]:  # bytes
+
+            with temppathlib.TemporaryDirectory() as tmp_dir:
+                for index in range(number_of_files):
+                    file = tmp_dir.path / "file{}".format(index)
+                    file.write_text("a" * size)
+                try:
+                    client = gswrap.Client(bucket_name=self.bucket)
+                    gswrap_start = time.time()
+                    self._gswrap_upload_files(client=client, path=tmp_dir.path)
+                    gswrap_end = time.time()
+
+                finally:
+                    subprocess.check_call([
+                        "gsutil", "-m", "-q", "rm", "-r",
+                        "{}".format(self.url_prefix)
+                    ])
+
+                try:
+                    gsutil_start = time.time()
+                    self._gsutil_upload_files(path=tmp_dir.path)
+                    gsutil_end = time.time()
+                finally:
+                    subprocess.check_call([
+                        "gsutil", "-m", "-q", "rm", "-r",
+                        "{}".format(self.url_prefix)
+                    ])
+
+            print_benchmark(
+                benchmark="Benchmark upload 3 files with {} bytes".format(size),
+                time_gswrap=gswrap_end - gswrap_start,
+                time_gsutil=gsutil_end - gsutil_start)
+
+    def _gswrap_upload_files(self, client: gswrap.Client, path: pathlib.Path):
+        client.cp(
+            src=path, dst=self.url_prefix, recursive=True, multithreaded=True)
+
+    def _gsutil_upload_files(self, path: pathlib.Path):
+        subprocess.check_call([
+            "gsutil", "-m", "-q", "cp", "-r", "{}".format(path.as_posix()),
+            "{}".format(self.url_prefix)
+        ])
+
+    def benchmark_upload_many_to_many_filtered(self):
+        for testcase in [10, 100, 1000]:
+            with temppathlib.TemporaryDirectory() as tmp_dir:
+                mod = 0
+                for index in range(testcase):
+                    mod += 1
```

Comment:
if you want to keep this mod operation you should explain in a comment what it does